--- a/gas/config/tc-arm.c
+++ b/gas/config/tc-arm.c
@@ -885,6 +885,9 @@
 static inline int
 skip_past_char (char ** str, char c)
 {
+  /* PR gas/14987: Allow for whitespace before the expected character.  */
+  skip_whitespace (*str);
+  
   if (**str == c)
     {
       (*str)++;
@@ -5168,6 +5171,9 @@
       return PARSE_OPERAND_SUCCESS;
     }
 
+  /* PR gas/14887: Allow for whitespace after the opening bracket.  */
+  skip_whitespace (p);
+
   if ((reg = arm_reg_parse (&p, REG_TYPE_RN)) == FAIL)
     {
       inst.error = _(reg_expected_msgs[REG_TYPE_RN]);
--- a/gas/testsuite/gas/arm/neon-ldst-es.s
+++ b/gas/testsuite/gas/arm/neon-ldst-es.s
@@ -57,3 +57,7 @@
 	vld2.32 {q1},[r7]
 	vld4.32 {q1-q2},[r7]
 	vld4.32 {q14-q15},[r7]
+
+	@ PR 14987 and 14887: Allow for whitespace in the instruction.
+	vld1.32 { d1 [ ] } , [ r2 ] , r3
+
