diff -Naur binutils-2.20/gas/Makefile.am 0/gas/Makefile.am
--- a/gas/Makefile.am	2009-09-07 09:16:06.000000000 +0000
+++ b/gas/Makefile.am	2009-10-17 02:35:04.000000000 +0000
@@ -780,13 +780,13 @@
 against=stage2
 
 # This rule is derived from corresponding code in the Makefile.in for gcc.
-# The "tail +16c" is to bypass headers which may include timestamps or
+# The "tail -c +16" is to bypass headers which may include timestamps or
 # temporary assembly file names.
 comparison:
 	x=0 ; \
 	for file in *.o ; do \
-	  tail +16c ./$$file > tmp-foo1; \
-	  if tail +16c ${against}/$$file > tmp-foo2 2>/dev/null ; then \
+	  tail -c +16 ./$$file > tmp-foo1; \
+	  if tail -c +16 ${against}/$$file > tmp-foo2 2>/dev/null ; then \
 	    if cmp tmp-foo1 tmp-foo2 ; then \
 	      true ; \
 	    else \
diff -Naur binutils-2.20/gas/Makefile.in 0/gas/Makefile.in
--- a/gas/Makefile.in	2009-09-07 12:09:27.000000000 +0000
+++ b/gas/Makefile.in	2009-10-17 02:35:04.000000000 +0000
@@ -2648,13 +2648,13 @@
 	if [ -f stage3/as-new$(EXEEXT) -a ! -f stage3/as$(EXEEXT) ] ; then (cd stage3 ; ln -s as-new as$(EXEEXT)) ; fi
 
 # This rule is derived from corresponding code in the Makefile.in for gcc.
-# The "tail +16c" is to bypass headers which may include timestamps or
+# The "tail -c +16" is to bypass headers which may include timestamps or
 # temporary assembly file names.
 comparison:
 	x=0 ; \
 	for file in *.o ; do \
-	  tail +16c ./$$file > tmp-foo1; \
-	  if tail +16c ${against}/$$file > tmp-foo2 2>/dev/null ; then \
+	  tail -c +16 ./$$file > tmp-foo1; \
+	  if tail -c +16 ${against}/$$file > tmp-foo2 2>/dev/null ; then \
 	    if cmp tmp-foo1 tmp-foo2 ; then \
 	      true ; \
 	    else \
diff -Naur binutils-2.20/src-release 0/src-release
--- a/src-release	2008-06-18 23:20:43.000000000 +0000
+++ b/src-release	2009-10-17 02:35:04.000000000 +0000
@@ -71,7 +71,7 @@
 	elif grep AM_INIT_AUTOMAKE $(TOOL)/configure.in >/dev/null 2>&1; then \
 	  sed < $(TOOL)/configure.in -n 's/AM_INIT_AUTOMAKE[^,]*, *\([^)]*\))/\1/p'; \
 	elif test -f $(TOOL)/version.in; then \
-	  head -1 $(TOOL)/version.in; \
+	  head -n 1 $(TOOL)/version.in; \
 	elif grep VERSION $(TOOL)/Makefile.in > /dev/null 2>&1; then \
 	  sed < $(TOOL)/Makefile.in -n 's/^VERSION *= *//p'; \
 	else \
